pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BaseReduceService.java

```diff
@@ -0,0 +1,308 @@
+    _reduceExecutorService = Executors.newFixedThreadPool(numThreadsInExecutorService, reduceThreadFactory);
+  }
+
+  protected static void updateAlias(QueryContext queryContext, BrokerResponseNative brokerResponseNative) {
+    ResultTable resultTable = brokerResponseNative.getResultTable();
+    if (resultTable == null) {
+      return;
+    }
+    List<String> aliasList = queryContext.getAliasList();
+    if (aliasList.isEmpty()) {
+      return;
+    }
+
+    String[] columnNames = resultTable.getDataSchema().getColumnNames();
+    List<ExpressionContext> selectExpressions = getSelectExpressions(queryContext.getSelectExpressions());
+    int numSelectExpressions = selectExpressions.size();
+    // For query like `SELECT *`, we skip alias update.
+    if (columnNames.length != numSelectExpressions) {
+      return;
+    }
+    for (int i = 0; i < numSelectExpressions; i++) {
+      String alias = aliasList.get(i);
+      if (alias != null) {
+        columnNames[i] = alias;
+      }
+    }
+  }
+
+  protected static List<ExpressionContext> getSelectExpressions(List<ExpressionContext> selectExpressions) {
+    // NOTE: For DISTINCT queries, need to extract the arguments as the SELECT expressions
+    if (selectExpressions.size() == 1 && selectExpressions.get(0).getType() == ExpressionContext.Type.FUNCTION
+        && selectExpressions.get(0).getFunction().getFunctionName().equals("distinct")) {
+      return selectExpressions.get(0).getFunction().getArguments();
+    }
+    return selectExpressions;
+  }
+
+  protected static void updateBrokerMetrics(BrokerResponseNative brokerResponseNative, Map<String, String> metadata) {
+
+  }
+
+  protected void shutDown() {
+    _reduceExecutorService.shutdownNow();
+  }
+
+  protected static class BrokerMetricsAggregator {
+    List<QueryProcessingException> _processingExceptions = new ArrayList<>();
+    Map<String, String> _traceInfo = new HashMap<>();
+    long _numDocsScanned = 0L;
+    long _numEntriesScannedInFilter = 0L;
+    long _numEntriesScannedPostFilter = 0L;
+    long _numSegmentsQueried = 0L;
+    long _numSegmentsProcessed = 0L;
+    long _numSegmentsMatched = 0L;
+    long _numConsumingSegmentsProcessed = 0L;
+    long _minConsumingFreshnessTimeMs = Long.MAX_VALUE;
+    long _numTotalDocs = 0L;
+    long _offlineThreadCpuTimeNs = 0L;
+    long _realtimeThreadCpuTimeNs = 0L;
+    long _offlineSystemActivitiesCpuTimeNs = 0L;
+    long _realtimeSystemActivitiesCpuTimeNs = 0L;
+    long _offlineResponseSerializationCpuTimeNs = 0L;
+    long _realtimeResponseSerializationCpuTimeNs = 0L;
+    long _offlineTotalCpuTimeNs = 0L;
+    long _realtimeTotalCpuTimeNs = 0L;
+    boolean _numGroupsLimitReached = false;
+    boolean _enableTrace;
+
+    protected BrokerMetricsAggregator(boolean enableTrace) {
+      _enableTrace = enableTrace;
+    }
+
+    protected void addMetrics(ServerRoutingInstance routingInstance, DataTable dataTable) {
```

Comment:
This needs to be thread safe, and suggest renaming it
```suggestion
    protected synchronized void aggregate(ServerRoutingInstance routingInstance, DataTable dataTable) {
```